src/Confluent.Kafka/Producer.cs

```diff
@@ -665,53 +666,96 @@ internal Producer(ProducerBuilder<TKey, TValue> builder)
+                    : asyncKeySerializer.SerializeAsync(message.Key, true, message, topicPartition)
+                        .ConfigureAwait(continueOnCapturedContext: false)
+                        .GetAwaiter()
+                        .GetResult();
+            }
+            catch (Exception exception)
+            {
+                throw new ProduceException<TKey, TValue>(
+                    new Error(ErrorCode.Local_KeySerialization),
+                    new DeliveryResult<TKey, TValue>
+                    {
+                        Message = message,
+                        TopicPartitionOffset = new TopicPartitionOffset(topicPartition, Offset.Invalid)
+                    },
+                    exception);
+            }
 
-            if (enableDeliveryReports)
+            byte[] valBytes;
+            try
             {
-                var handler = new TypedTaskDeliveryHandlerShim<TKey, TValue>(topicPartition.Topic,
-                    enableDeliveryReportKey ? message.Key : default(TKey),
-                    enableDeliveryReportValue ? message.Value : default(TValue));
+                valBytes = (valueSerializer != null)
+                    ? valueSerializer.Serialize(message.Value, false, message, topicPartition)
+                    : asyncValueSerializer.SerializeAsync(message.Value, false, message, topicPartition)
+                        .ConfigureAwait(continueOnCapturedContext: false)
+                        .GetAwaiter()
+                        .GetResult();
+            }
+            catch (Exception exception)
+            {
+                throw new ProduceException<TKey, TValue>(
+                    new Error(ErrorCode.Local_ValueSerialization),
+                    new DeliveryResult<TKey, TValue>
+                    {
+                        Message = message,
+                        TopicPartitionOffset = new TopicPartitionOffset(topicPartition, Offset.Invalid)
+                    },
+                    exception);
+            }
 
-                cancellationToken.Register(() => handler.TrySetException(new TaskCanceledException()));
+            try
+            {
+                if (enableDeliveryReports)
+                {
+                    var handler = new TypedTaskDeliveryHandlerShim<TKey, TValue>(topicPartition.Topic,
+                        enableDeliveryReportKey ? message.Key : default(TKey),
+                        enableDeliveryReportValue ? message.Value : default(TValue));
 
-                ProduceImpl(
-                    topicPartition.Topic,
-                    valBytes, 0, valBytes == null ? 0 : valBytes.Length,
-                    keyBytes, 0, keyBytes == null ? 0 : keyBytes.Length,
-                    message.Timestamp, topicPartition.Partition, message.Headers,
-                    handler);
+                    cancellationToken.Register(() => handler.TrySetException(new TaskCanceledException()));
```

Comment:
don't worry about this. it doesn't work and it's going away in the future clean-up-the-cancellationtoken-stuff PR.